### PR TITLE
feat: implemented native authentication with pwa using auth0 API

### PR DESCRIPTION
**Pull request by vickywane**
## Purpose
This PR implements a native authentication with a user's PWA, as a requirement for the premium user [feature](https://github.com/ambianic/ambianic-subscriptions/issues/1) 

## Approach
Work done here fetches the user's access_token from auth0 after the user's consent from the PWA.

[This](https://drive.google.com/file/d/1XTnYAcLO32OouO30iK953vqiR7ck67Kz/view?usp=sharing) design diagram shows how the notification feature is being implemented in this Pull Request and also how it integrates with the connected PWA. 

edge-device-pwa-sync-flow (5)

## Merge Checklist
- [x]  The code change is tested and works locally.
- [x]  The user and dev documentation is up to date.
- [x]  There is no commented-out code in this PR.
- [x]  No lint errors (use flake8)
- [ ]  100% test coverage

### CODE REVIEW

**Comment by ivelin** [comment]
@vickywane the CI checks are failing. Let me know when all checks pass.

I would also like to review simultaneously the UI PR that handles user authorization approval for this edge device.

File: src/ambianic/pipeline/store.py [hunk 1]
```diff
@@ -174,16 +177,35 @@ def notify(self, save_json: dict):
...
         for inference_result in save_json['inference_result']:
-            data = {
-                'id': save_json['id'],
-                'label': inference_result['label'],
-                'confidence': inference_result['confidence'],
-                'datetime': save_json['datetime'],
-            }
-            notification = Notification(data=data, providers=self.notification_config["providers"])
-            self.notification.send(notification)
+            try:
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:184]: @vickywane why hardcode the assumption that the edge device needs premium subscription. What if the user does not want to have premium subscription at all. This code should only trigger if the user has elected to sign up for premium services and consequently the PWA has informed the edge devices about that. Otherwise this code should not be part of the edge flow at all.

Comment (ivelin) [reply src/ambianic/pipeline/store.py:184]: In fact I think we should move to a completely independent local library all the logic related to turning on premium subscriptions. It will be likely used for functions other than notifications in the future.

File: src/ambianic/pipeline/store.py [hunk 2]
```diff
@@ -174,16 +177,35 @@ def notify(self, save_json: dict):
...
+                    credentials = file_data['credentials']
+
+                    header = {
+                        'Authorization': 'Bearer {0}'.format(
+                            credentials['ACCESS_TOKEN'])}
+
+                    userRequest = get(
+                        '{0}/userinfo'.format(credentials['AUTH_DOMAIN']), header)
+                    userData = userRequest.json()
+                    # TODO make a request to the serverless function using user's data
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:196]: Do we need to implement in code the call out to a premium rest notification service? Can we re-use the already available apprise plugin with the configuration settings from config.yaml for the specific service?
The PWA can ask the user for the notification settings (email subject, whether to include an attachment, etc) push these via the /config edge api. Then the notification service will do its job. That's already implemented.

In other words, I am not sure we need to modify `store.py`.

File: src/ambianic/pipeline/store.py [hunk 3]
```diff
@@ -174,16 +177,35 @@ def notify(self, save_json: dict):
...
+
         # TODO extract inference data
         for inference_result in save_json['inference_result']:
-            data = {
-                'id': save_json['id'],
-                'label': inference_result['label'],
-                'confidence': inference_result['confidence'],
-                'datetime': save_json['datetime'],
-            }
-            notification = Notification(data=data, providers=self.notification_config["providers"])
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:185]: @vickywane a critical feature we need to consider for the notification service is potential failure to call the rest api due to internet disconnect or cloud service downtime. In such cases we need to have some kind of local queue for outgoing notifications up to size N and periodically retry to push them out.

Maybe this feature can be added to the edge notification service in general and it will benefit all notifications , not just premium.

File: src/ambianic/pipeline/store.py [hunk 4]
```diff
@@ -174,16 +177,35 @@ def notify(self, save_json: dict):
...
-            self.notification.send(notification)
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:186]: @vickywane I was referring earlier to this line. We probably need to upgrade the implementation of send() so it handles failures and schedules retries. Maybe we can do that in a separate issue and PR. WDYT?

We also need to think about throttling. In case somethings goes wrong and too many notification events are somehow pushed through the notification service. We should consider this possibly as a separate issue and PR.

File: src/ambianic/webapp/flaskr.py [hunk 5]
```diff
@@ -159,6 +161,66 @@ def get_status():
         resp = jsonify(response_object)
         return resp
 
+    # start the auth0 authentication process with creating the auth0 keys
```
Comment (ivelin) [inline src/ambianic/webapp/flaskr.py:164]: @vickywane these api token management rest api methods make sense, but I need to see them in a more detailed UML sequence diagram to understand exactly how they will fit in the PWA UX flow.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane the codecov check does not pass. This PR doesn't have a single test to cover the new code.

See other comments inline.

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:186]: Yes, these are more advanced use cases. 

They would be implemented as this feature matures.

File: src/ambianic/webapp/flaskr.py (hunk 5 above)
Comment (vickywane) [reply src/ambianic/webapp/flaskr.py:164]: The existing UML diagram has been updated and referenced on Slack

File: src/ambianic/pipeline/store.py (hunk 3 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:185]: Can we use a queuing service for this? 

Or what are your recommendations?

File: src/ambianic/pipeline/store.py (hunk 2 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:196]: Okay. 

I am looking into using Aprise notfication service

File: src/ambianic/webapp/flaskr.py (hunk 5 above)
Comment (ivelin) [reply src/ambianic/webapp/flaskr.py:164]: It needs to be referenced here, in the issue that it solves and most importantly in the project DOCS!! Slack is not a searchable web resource. It's only useful for short lived chats.

Comment (vickywane) [reply src/ambianic/webapp/flaskr.py:164]: I have updated this Pull request description to contain the document link for better visibility. [This](https://app.diagrams.net/#G1gDYMs2LjbZ9dnJG8i5yl5UFMGAJ1nABq) is also the link.

The `readme` within this PR would also be updated to include the new Notifications

**Comment by ivelin** [comment]
> [This](https://app.diagrams.net/#G1gDYMs2LjbZ9dnJG8i5yl5UFMGAJ1nABq) design diagram shows how the notification feature is being implemented in this Pull Request and also how it integrates with the connected PWA.

I am not able to open this diagram.

Screen Shot 2021-04-02 at 5 56 42 PM

**Comment by vickywane** [comment]
> > [This](https://app.diagrams.net/#G1gDYMs2LjbZ9dnJG8i5yl5UFMGAJ1nABq) design diagram shows how the notification feature is being implemented in this Pull Request and also how it integrates with the connected PWA.
> 
> @vickywane
> I am not able to open this diagram. The URL points to an error.
> 
 Seems to be an issue with [Draw](https://draw.io). I just duplicated the diagram in a new document [here](https://drive.google.com/file/d/1TW_GV5xaDzwuSrjdSJBaPprLUsew9TsM/view?usp=sharing)

**Comment by ivelin** [comment]
> > > [This](https://app.diagrams.net/#G1gDYMs2LjbZ9dnJG8i5yl5UFMGAJ1nABq) design diagram shows how the notification feature is being implemented in this Pull Request and also how it integrates with the connected PWA.
> > 
> > 
> > @vickywane
> > I am not able to open this diagram. The URL points to an error.
> 
> Seems to be an issue with [Draw](https://draw.io). I just duplicated the diagram in a new document [here](https://drive.google.com/file/d/1TW_GV5xaDzwuSrjdSJBaPprLUsew9TsM/view?usp=sharing)

OK. Added google doc comments.
Screen Shot 2021-04-02 at 6 21 35 PM

**Comment by vickywane** [comment]
I am currently writing tests for the new API endpoints introduced as part of the edge-PWA auth flow. 

The first endpoint (`/user-code`)has been covered with tests.
 
However, there's seems to be a limit to what can be tested. The `/verify-token` route returns a URL to a page containing a button that has to be manually clicked by a user before the authentication is complete. This cannot be done within the test suite except using simulation tests.

This also affects the third endpoint (`/save-token`). The authentication has be complete before a valid token is returned by auth0.

File: src/ambianic/pipeline/store.py (hunk 1 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:184]: Okay. 

Can this come in a follow-up PR when we have more use cases for the premium subscription?

Comment (vickywane) [reply src/ambianic/pipeline/store.py:184]: For now, the `store.py` has been readjusted to accommodate non-premium users.

As illustrated in the design diagram, obtaining the `ACCESS_TOKEN` is the last step in the premium notification flow, hence, the premium notification flow is only used when there is an `ACCESS_TOKEN` field in the `premium.yaml` file.

Comment (vickywane) [reply src/ambianic/pipeline/store.py:184]: > @vickywane why hardcode the assumption that the edge device needs premium subscription. What if the user does not want to have premium subscription at all. This code should only trigger if the user has elected to sign up for premium services and consequently the PWA has informed the edge devices about that. Otherwise this code should not be part of the edge flow at all.

The premium notification flow has been re-adjusted to fix this concern. Rather than re-implementing a new notification flow with `aprise`, the premium notification flow would add the needed credentials to the `config.yaml` file so it can be used by the existing `NotificationHandler`, to avoid duplicating code.

File: src/ambianic/webapp/flaskr.py (hunk 5 above)
Comment (ivelin) [reply src/ambianic/webapp/flaskr.py:164]: Link leads to inaccessible
Screen Shot 2021-04-05 at 8 44 00 AM
 file.

Comment (ivelin) [reply src/ambianic/webapp/flaskr.py:164]: Make sure to embed the diagram view in the comments AND verify access with anonymous browser to save us back and forth on these access issues.

File: src/ambianic/webapp/flaskr.py [hunk 6]
```diff
@@ -159,6 +161,99 @@ def get_status():
...
+    @app.route('/api/auth/user-code', methods=['POST'])
```
Comment (ivelin) [inline src/ambianic/webapp/flaskr.py:165]: Too much code in the main flaskr file. Move auth management in its own python module with its own unit tests.

File: src/ambianic/webapp/flaskr.py [hunk 7]
```diff
@@ -159,6 +161,99 @@ def get_status():
...
+                data={
+                    "client_id": "{0}".format(
+                        credentials['CLIENT_ID']),
+                    "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+                    "device_code": device_code})
+            response = req.json()
+
+            return response
+
+    @app.route('/api/auth/save-token', methods=['POST'])
```
Comment (ivelin) [inline src/ambianic/webapp/flaskr.py:212]: All this new code needs to be covered with unit tests.

**Review by ivelin — CHANGES_REQUESTED** [review]
> I am currently writing tests for the new API endpoints introduced as part of the edge-PWA auth flow.
> 
> The first endpoint (`/user-code`)has been covered with tests.
> 
> However, there's seems to be a limit to what can be tested. The `/verify-token` route returns a URL to a page containing a button that has to be manually clicked by a user before the authentication is complete. This cannot be done within the test suite except using simulation tests.
> 
> This also affects the third endpoint (`/save-token`). The authentication has be complete before a valid token is returned by auth0.

Unit tests should not depend on third party services. Isolate and test your own code. 
Worry about integration tests after you have full unit test coverage.

**Comment by ivelin** [comment]
@vickywane read a bit more how to create proper activity design diagrams. You have options, but whichever one you pick must communicate clearly who and when does what exactly:
https://drawio-app.com/create-uml-sequence-diagrams-in-draw-io/
https://www.diagrams.net/blog/swimlane-diagrams
https://nathanhumphrey.ca/posts/2021-01-11-bpmn-with-diagrams-net/

**Comment by vickywane** [comment]
> @vickywane read a bit more how to create proper activity design diagrams. You have options, but whichever one you pick must communicate clearly who and when does what exactly:
> https://drawio-app.com/create-uml-sequence-diagrams-in-draw-io/
> https://www.diagrams.net/blog/swimlane-diagrams
> https://nathanhumphrey.ca/posts/2021-01-11-bpmn-with-diagrams-net/

Thank you. 

I just expanded the document following guides from the articles above. The updated image is in the PR description above.

**Comment by ivelin** [comment]
> > @vickywane read a bit more how to create proper activity design diagrams. You have options, but whichever one you pick must communicate clearly who and when does what exactly:
> > https://drawio-app.com/create-uml-sequence-diagrams-in-draw-io/
> > https://www.diagrams.net/blog/swimlane-diagrams
> > https://nathanhumphrey.ca/posts/2021-01-11-bpmn-with-diagrams-net/
> 
> Thank you.
> 
> I just expanded the document following guides from the articles above. The updated image is in the PR description above.

The link in the PR description text points to an outdated document. Make sure the view and the link are in sync.  Even better, make the PNG clickable and linking directly to the latest diagram source so it's easy to collaborate on it.

**Comment by vickywane** [comment]
> > > @vickywane read a bit more how to create proper activity design diagrams. You have options, but whichever one you pick must communicate clearly who and when does what exactly:
> > > https://drawio-app.com/create-uml-sequence-diagrams-in-draw-io/
> > > https://www.diagrams.net/blog/swimlane-diagrams
> > > https://nathanhumphrey.ca/posts/2021-01-11-bpmn-with-diagrams-net/
> > 
> > 
> > Thank you.
> > I just expanded the document following guides from the articles above. The updated image is in the PR description above.
> 
> The link in the PR description text points to an outdated document. Make sure the view and the link are in sync. Even better, make the PNG clickable and linking directly to the latest diagram source so it's easy to collaborate on it.

The link has been fixed. It was becoming large so I had to use another means to export.

File: src/ambianic/webapp/flaskr.py (hunk 6 above)
Comment (vickywane) [reply src/ambianic/webapp/flaskr.py:166]: As agreed, the premium auth management would be split into its own module in a follow-up PR. 

They are currently being covered with unit tests.

**Comment by ivelin** [comment]
> > The link in the PR description text points to an outdated document. Make sure the view and the link are in sync. Even better, make the PNG clickable and linking directly to the latest diagram source so it's easy to collaborate on it.
> 
> The link has been fixed. It was becoming large so I had to use another means to export.

The link now points to a google doc, that does not allow comments....

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (ivelin) [reply src/ambianic/pipeline/store.py]: OK, open issues to track these problems separately.

File: src/ambianic/webapp/flaskr.py (hunk 6 above)
Comment (ivelin) [reply src/ambianic/webapp/flaskr.py:165]: Create an issue to make sure this discussion does not get lost.

**Comment by vickywane** [comment]
> The link now points to a google doc, that does not allow comments....

I have updated the file permission [here](https://drive.google.com/file/d/1TW_GV5xaDzwuSrjdSJBaPprLUsew9TsM/view?usp=sharing).

**Comment by ivelin** [comment]
> > The link now points to a google doc, that does not allow comments....
> 
> I have updated the file permission [here](https://drive.google.com/file/d/1TW_GV5xaDzwuSrjdSJBaPprLUsew9TsM/view?usp=sharing).

OK, I can comment now.

File: config.yaml [hunk 8]
```diff
@@ -33,47 +33,47 @@ timeline:
...
   recorded_cam_feed:
     uri: file:///workspace/tests/pipeline/avsource/test2-cam-person1.mkv
-#    type: video
+    type: video
 #    live: true
 
 ai_models:
   image_detection:
     model:
-      tflite: /opt/ambianic-edge/ai_models/mobilenet_ssd_v2_coco_quant_postprocess.tflite
```
Comment (ivelin) [inline config.yaml:56]: @vickywane the path depends on how the code is installed. Instead of changing the default config, the correct approach is to use the correct setup scripts for different environments. We discussed on the roundtable call. See how we build rpi OS image and similarly create a separate setup script for dev simulation of edge install.

File: docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
+version: "3.7"
```
Comment (ivelin) [inline docker-compose.yml:1]: @vickywane this docker-compose recipe should be part of a setup for dev install. Either a sub-directory of this repo or a separate repo.

File: src/ambianic/pipeline/store.py (hunk 3 above)
Comment (ivelin) [reply src/ambianic/pipeline/store.py:185]: @vickywane this is still an open topic. Either address in this PR or open a separate issue.

File: src/ambianic/pipeline/store.py [hunk 10]
```diff
@@ -171,18 +175,52 @@ def process_sample(self, **sample) -> Iterable[dict]:
...
+                                'notification': {
+                                    'title': 'Ambianic.ai New {0} event'.format(
+                                        data['label']),
+                                    'message': 'New {0} detected with a {1} confidence level'.format(
+                                        data['label'],
+                                        data['confidence']),
+                                }})
+
+            except FileNotFoundError:
+                log.info("Error locating file")
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:211]: @vickywane how would the user know if they missed a notification? Local logging is good for debugging, but how are we going to inform the user that edge is unable to send notifications?

File: src/ambianic/pipeline/store.py [hunk 11]
```diff
@@ -171,18 +175,52 @@ def process_sample(self, **sample) -> Iterable[dict]:
...
+
+            except TypeError:
+                log.info("Auth0 key not found")
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:214]: @vickywane errors are usually logged with at least warn or error level, not info.

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (ivelin) [reply src/ambianic/pipeline/store.py:186]: @vickywane how did you address this open thread?

File: src/ambianic/webapp/flaskr.py [hunk 12]
```diff
@@ -159,6 +162,29 @@ def get_status():
...
+
+        directory = pkg_resources.resource_filename(
+            "ambianic.webapp", "premium.yaml")
+
+        try:
+            file = open(directory, 'w+')
+            yaml.dump(auth_file, file)
+            file.close()
+        except FileNotFoundError:
+            log.debug("Error locating file")
```
Comment (ivelin) [inline src/ambianic/webapp/flaskr.py:184]: @vickywane same comment here. Do we need a warn or error log level? Is debug sufficient?

File: src/ambianic/pipeline/store.py [hunk 13]
```diff
@@ -171,18 +175,52 @@ def process_sample(self, **sample) -> Iterable[dict]:
...
             }
-            notification = Notification(data=data, providers=self.notification_config["providers"])
+
+            try:
+                premiumFile = pkg_resources.resource_filename(
+                    "ambianic.webapp", "premium.yaml")
+                with open(premiumFile, "r") as file:
+                    configFile = yaml.safe_load(file)
+                    userId = configFile['credentials']["USER_AUTH0_ID"]
+                    endpoint = configFile['credentials']["NOTIFICATION_ENDPOINT"]
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:196]: @vickywane I can't find the code or configuration that calls out the notification API endpoint.
I was expecting to see a place where the pipeline config is updated to send notifications via the premium API. How else would apprise know the credentials and endpoint where to direct event notifications.

File: config.yaml [hunk 14]
```diff
@@ -33,47 +33,47 @@ timeline:
 # Using Home Assistant conventions to ease upcoming integration
 sources:
 
-  # direct support for raspberry picamera
```
[the remark this answers is not in the document]
Comment (vickywane) [reply config.yaml:36]: This was actually refactored by the python `yaml` package. Each time it writes into the file, it clears out commented code and reformats the file. 

I am currently looking for ways to configure it so it doesn't do this.

File: config.yaml (hunk 8 above)
Comment (vickywane) [reply config.yaml:56]: I would take note of this.

File: src/ambianic/pipeline/store.py (hunk 10 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:211]: Logging of the sent notifications is an advanced feature which I would treat in a separate pull request. 

On the PWA side, I plan to refactor the settings page to contain sections about the notification and subscription feature. Logging from the edge device would be among the item it contains.

File: src/ambianic/pipeline/store.py (hunk 11 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:214]: Noted. I would correct that. 

However, in development, I can only see items logged using `log.info()`

File: src/ambianic/pipeline/store.py (hunk 13 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:196]: The configuration is stored in a `premium.yaml` file, which is created when the PWA syncs with the edge device. 

As shown in the snippet above, the credentials are gotten from the PWA through the `premium-notifcation` endpoint [here](https://github.com/vickywane/ambianic-edge/blob/auth/auth0-pwa/src/ambianic/webapp/flaskr.py#L165).
[See](https://github.com/ambianic/ambianic-ui/pull/562#discussion_r628458493) my response to this question on the PWA pull request.

You can also refer to the design document for more clarification as this process is documented there.

File: src/ambianic/webapp/flaskr.py (hunk 12 above)
Comment (vickywane) [reply src/ambianic/webapp/flaskr.py:184]: I think `warn` is sufficent. 

I however can see the `debug` level being used in other parts of the codebase.

Comment (vickywane) [reply src/ambianic/webapp/flaskr.py:184]: >  how did you address this open thread?

Are you making reference to the file opened using the `with` statement?

If yes, 
It is automatically closed and I do not need to manually close it as I using the `with` statement not `open`.

[This](https://lerner.co.il/2015/01/18/dont-use-python-close-files-answer-depends/) article contains more explanation on this.

File: docker-compose.yml (hunk 9 above)
Comment (vickywane) [reply docker-compose.yml:1]: @ivelin Going through the `docker-compose.yml` file, I can see there is no need for it. 

The existing `ambianic-start.sh` script is enough to run the edge device.

File: src/ambianic/pipeline/store.py [hunk 15]
```diff
@@ -171,18 +175,52 @@ def process_sample(self, **sample) -> Iterable[dict]:
...
-            notification = Notification(data=data, providers=self.notification_config["providers"])
+
+            try:
+                premiumFile = pkg_resources.resource_filename(
+                    "ambianic.webapp", "premium.yaml")
+                with open(premiumFile, "r+") as file:
+                    configFile = yaml.safe_load(file)
+                    userId = configFile['credentials']["USER_AUTH0_ID"]
+                    endpoint = configFile['credentials']["NOTIFICATION_ENDPOINT"]
+                    if (userId):
```
Comment (ivelin) [inline src/ambianic/pipeline/store.py:197]: This section with premium notification logic should be in the `ambianic.notification` module , not hardcoded in `store.py`.

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:186]: It is automatically closed and I do not need to manually close it as I am using the [with](https://www.geeksforgeeks.org/with-statement-in-python/) statement not `open`.

[This](https://lerner.co.il/2015/01/18/dont-use-python-close-files-answer-depends/) article contains more explanation on this.

File: src/ambianic/pipeline/store.py (hunk 15 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py]: Alright. This adjustment has been made. 

I created a function ( isolated from the `Notification` class ) in the `Notification` file to handle sending the emails using the Cloud API.

**Comment by vickywane** [comment]
> still a few open comments that are awaiting response

I think I have responded and made adjustments to all raised comments.
Please mention me on any comment that has been skipped or not answered fully.

**Comment by ivelin** [comment]
> Hello @vickywane! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> 
> * In the file [`src/ambianic/notification.py`](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/src/ambianic/notification.py):
> 
> > [Line 35:80](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/src/ambianic/notification.py#L35): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> 
> * In the file [`src/ambianic/pipeline/store.py`](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/src/ambianic/pipeline/store.py):
> 
> > [Line 11:80](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/src/ambianic/pipeline/store.py#L11): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> > [Line 72:80](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/src/ambianic/pipeline/store.py#L72): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> 
> * In the file [`tests/test_flaskr.py`](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/tests/test_flaskr.py):
> 
> > [Line 77:80](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/tests/test_flaskr.py#L77): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> > [Line 81:80](https://github.com/ambianic/ambianic-edge/blob/f6a0d0cc9dfff79bc16af754f175c8b5e97de34f/tests/test_flaskr.py#L81): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> 
> ##### Comment last updated at 2021-05-12 19:16:16 UTC

@vickywane please address these linting issues.

File: src/ambianic/pipeline/store.py (hunk 3 above)
Comment (ivelin) [reply src/ambianic/pipeline/store.py:185]: @vickywane this is still an open topic...

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (ivelin) [reply src/ambianic/pipeline/store.py:186]: @vickywane `with` only handles automatic file close. It does not address the issue of retrying a message if it failed to send, nor does it address throttling (i.e. sending too many messages to the API that cause denial of service).

**Comment by ivelin** [comment]
> I am currently writing tests for the new API endpoints introduced as part of the edge-PWA auth flow.
> 
> The first endpoint (`/user-code`)has been covered with tests.
> 
> However, there's seems to be a limit to what can be tested. The `/verify-token` route returns a URL to a page containing a button that has to be manually clicked by a user before the authentication is complete. This cannot be done within the test suite except using simulation tests.
> 
> This also affects the third endpoint (`/save-token`). The authentication has be complete before a valid token is returned by auth0.

@vickywane use mockups to emulate server side behavior.

https://changhsinlee.com/pytest-mock/
https://www.toptal.com/python/an-introduction-to-mocking-in-python

**Comment by vickywane** [comment]
> > I am currently writing tests for the new API endpoints introduced as part of the edge-PWA auth flow.
> 
> > 
> 
> > The first endpoint (`/user-code`)has been covered with tests.
> 
> > 
> 
> > However, there's seems to be a limit to what can be tested. The `/verify-token` route returns a URL to a page containing a button that has to be manually clicked by a user before the authentication is complete. This cannot be done within the test suite except using simulation tests.
> 
> > 
> 
> > This also affects the third endpoint (`/save-token`). The authentication has be complete before a valid token is returned by auth0.
> 
> 
> 
> @vickywane use mockups to emulate server side behavior.
> 
> 
> 
> https://changhsinlee.com/pytest-mock/
> 
> https://www.toptal.com/python/an-introduction-to-mocking-in-python
> 
> 

This is an old and outdated comment. It was part of the previous implementation which has been discarded. 

We no longer have those routes / endpoints in the codebase.

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:186]: Yes  it doesnt. 

Those are quite advanced use cases. We have [this](https://github.com/ambianic/ambianic-subscriptions/issues/11) open issue within the `ambianic-subscription` repository to handle such cases.

Also, implementing throttling and email retries may likely be implemented on the Cloud API where the email is sent from using our third-party email provider.

File: src/ambianic/pipeline/store.py (hunk 10 above)
Comment (ivelin) [reply src/ambianic/pipeline/store.py]: @vickywane 
Is there an open issue for this future work on proper error handling when notifications fail to be sent?

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (ivelin) [reply src/ambianic/pipeline/store.py:186]: OK, open new separate issues for each of these scenarios and link them back to this discussion.

File: src/ambianic/notification.py [hunk 16]
```diff
@@ -4,13 +4,52 @@
...
+import pkg_resources
+import yaml
+from requests import post
 
 log = logging.getLogger(__name__)
 
 UI_BASEURL = "https://ui.ambianic.ai"
 
+
+def sendCloudNotification(data):
```
Comment (ivelin) [inline src/ambianic/notification.py:16]: @vickywane `sendCloudNotification` is a redundant function. The existing notification class via apprise already allows json notification provider. You just need to pass to it the correct Ambianic Cloud API credentials and URL.
https://github.com/ambianic/ambianic-edge/blob/0291b2abfd76fd9a985a5b09445a505dfd4f83ca/tests/pipeline/test_notify.py#L135

**Review by ivelin — CHANGES_REQUESTED** [review]
Cleanup redundant notification code and address unresolved comments.

File: src/ambianic/notification.py (hunk 16 above)
Comment (vickywane) [reply src/ambianic/notification.py:16]: One of the requirements of this feature is to grant the least privilege to the edge device by supplying only the credentials needed. 

I am intentionally not utilizing Apprise because I do not want to pass the Mailgun credentials to the Edge device. Currently, the `sendNotification` function simply makes an HTTP request directly to the Cloud API, passing the user's ID and the event detection.  

It is the Cloud API's function to send the email using Mailgun client libraries. 

The existing Apprise configuration would still be valid for users who want to manually configure their device notification.

We have discussed this flow previously and it is also documented in the design diagram attached to this pull request.

File: src/ambianic/pipeline/store.py (hunk 10 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py]: Yes, [this issue](https://github.com/ambianic/ambianic-subscriptions/issues/12) has been opened and its implementation would begin when this pull request is merged.

File: src/ambianic/pipeline/store.py (hunk 3 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:185]: [This issue](https://github.com/ambianic/ambianic-subscriptions/issues/12) has been opened in the `Ambianic-subscriptions` repository  to report the need for notification retries at intervals

File: src/ambianic/webapp/flaskr.py (hunk 6 above)
Comment (vickywane) [reply src/ambianic/webapp/flaskr.py]: [This](https://github.com/ambianic/ambianic-edge/issues/352) issue has been created to handle this concern.

File: src/ambianic/pipeline/store.py (hunk 4 above)
Comment (vickywane) [reply src/ambianic/pipeline/store.py:186]: Okay. The respective issues have been opened and linked to the comments.

File: src/ambianic/notification.py (hunk 16 above)
Comment (vickywane) [reply src/ambianic/notification.py:16]: @ivelin [this](https://github.com/ambianic/ambianic-edge/issues/353) issue has been created as a follow-up to handle the adjustment of the existing `Notification`  class.

**Comment by vickywane** [comment]
As requested I have created the following issues to be immediately worked upon after this pull request is merged;
 
- Splitting auth management code into a separate module #352 
- Rewrite NotificationHandler to accommodate sending premium notification #353 

The following were created in the Cloud API repository as they would be implemented there;

- [Implement retries of undelivered notification as set interval](https://github.com/ambianic/ambianic-subscriptions/issues/12)
- [Implement user subscription credit budget](https://github.com/ambianic/ambianic-subscriptions/issues/11)

**Review by ivelin — APPROVED** [review]
@vickywane looks good. merging.
